/*
  The server is an SLM server which can display greyscale images on a fullscreen background.

  The server can accept a binarised uint8 numpy array (0-255) colours which it then displays.
  The position of the image on the screen can also be changed.
*/

syntax = "proto3";

package slm;

message ImageDescription {
  uint32 width = 1;
  uint32 height = 2;
  enum ColourType {
    GREY8 = 0;
    RGB8 = 1;
  };
  ColourType colour_type = 3;
}

message ImageData {
  oneof image_oneof {
    ImageDescription description = 11;
    bytes data = 4;
  }
}

message ScreenReply {
  int32 num_screens = 5;
}

message Screen {
  int32 screen = 6;
}

message Position {
  int32 x = 7;
  int32 y = 8;
}

message EmptyParams {}

message Response {
  bool completed = 9;
  string error = 10;
}

service SLM {
  // Set the image from a given image description and a stream of the image bytes
  rpc SetImage(stream ImageData) returns (Response) {}
  // Set the screen the slm is appearing on
  rpc SetScreen(Screen) returns (Response) {}
  // Set the position on the screen
  rpc SetPosition(Position) returns (Response) {}
}